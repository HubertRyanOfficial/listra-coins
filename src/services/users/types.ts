export type UserPostType = {
  email: string;
  password: string;
  name: string;
  balance: number;
  created_at: number;
  profileImage: string | null;
};

export type UserGetType = UserPostType & {
  id: string;
};
