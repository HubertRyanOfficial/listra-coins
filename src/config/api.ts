import axios from "axios";
import * as Device from "expo-device";

const api = axios.create({
  baseURL: !Device.isDevice
    ? "http://localhost:3000"
    : "http://192.168.0.104:3000",
});

export default api;
