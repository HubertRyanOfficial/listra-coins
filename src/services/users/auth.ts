import api from "@/config/api";
import { UserGetType } from "./types";

export const loginUser = async (email: string, password: string) => {
  const response = await api.get<UserGetType[]>("/users");
  const users = response.data;

  const currentUser = users.find((user) => user.email === email);

  if (currentUser) {
    if (currentUser.password === password) {
      return {
        id: currentUser.id,
        email: currentUser.email,
        name: currentUser.name,
        balance: currentUser.balance,
        created_at: currentUser.created_at,
      };
    } else {
      throw "incorrect-password";
    }
  } else {
    throw "user-not-exists";
  }
};
