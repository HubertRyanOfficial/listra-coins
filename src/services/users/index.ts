import api from "@/config/api";
import { UserGetType, UserPostType } from "./types";

export const createUser = async (data: UserPostType) => {
  const response = await api.get<UserGetType[]>("/users");

  const getCurrentUser = response.data.find(
    (user) => user.email === data.email
  );

  if (getCurrentUser) {
    throw "user-already-exists";
  }

  const user = await api.post("/users", data);
  return user.data;
};

export const updateUser = async (data: Partial<UserPostType>, userId) => {
  const user = await api.get(`/users/${userId}`);
  const response = await api.put<UserGetType>(`/users/${userId}`, {
    ...user.data,
    ...data,
  });
  return response.data;
};
