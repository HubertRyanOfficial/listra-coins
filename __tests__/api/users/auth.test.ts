import { loginUser } from "@/services/users/auth";
import api from "@/config/api";
import MockAdapter from "axios-mock-adapter";

const mock = new MockAdapter(api);

const mockUsers = [
  {
    id: "1",
    email: "user1@example.com",
    password: "password1",
    name: "User 1",
    balance: 100,
    created_at: Date.now(),
    profileImage: null,
    notificationToken: null,
  },
  {
    id: "2",
    email: "user2@example.com",
    password: "password2",
    name: "User 2",
    balance: 200,
    created_at: Date.now(),
    profileImage: null,
    notificationToken: null,
  },
];

describe("loginUser", () => {
  test("logs in user with correct email and password", async () => {
    mock.onGet("/users").reply(200, mockUsers);

    const email = "user1@example.com";
    const password = "password1";
    const loggedInUser = await loginUser(email, password);

    expect(loggedInUser).toEqual({
      id: "1",
      email: "user1@example.com",
      name: "User 1",
      balance: 100,
      created_at: expect.any(Number),
      profileImage: null,
      notificationToken: null,
    });
  });

  test("throws error for incorrect password", async () => {
    mock.onGet("/users").reply(200, mockUsers);

    const email = "user1@example.com";
    const password = "wrongpassword";

    await expect(
      loginUser(email, password)
    ).rejects.toThrowErrorMatchingSnapshot("incorrect-password");
  });

  test("throws error for user not exists", async () => {
    mock.onGet("/users").reply(200, mockUsers);

    const email = "nonexistentuser@example.com";
    const password = "password";

    await expect(
      loginUser(email, password)
    ).rejects.toThrowErrorMatchingSnapshot("user-not-exists");
  });
});
