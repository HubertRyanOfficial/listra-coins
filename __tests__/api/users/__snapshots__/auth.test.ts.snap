// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loginUser throws error for incorrect password: incorrect-password 1`] = `undefined`;

exports[`loginUser throws error for user not exists: user-not-exists 1`] = `undefined`;
